### ISSUE TITLE
Isolate Stdlib Extension Modules

### ISSUE THREAD

**Issue by ericsnowcurrently** [python/cpython#103092]
See [PEP 687](https://peps.python.org/pep-0687/).

Currently most stdlib extension have been ported to multi-phase init.  There are still a number of them to be ported, almost entirely non-builtin modules.  Also, some that have already been ported still have global state that needs to be fixed.

(This is part of the effort to finish isolating multiple interpreters from each other.  See gh-100227.)

### High-Level Info

How to isolate modules: https://docs.python.org/3/howto/isolating-extensions.html (AKA PEP 630).

The full list of modules that need porting can be found with:  `...`

The full list of remaining (unsupported) global variables is:

* builtin extensions: https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L308-L335
* non-builtin extensions: https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L338-L551

A full analysis of the modules may be found at the bottom of this post.

<details>
<summary>(other info)</summary>

#### Previous Work
* https://github.com/python/cpython/issues/84258
* _xxsubinterpreters [[issue](https://github.com/python/cpython/issues/99741)] [[PR](https://github.com/python/cpython/pull/99742)]
* _testinternalcapi [[issue](https://github.com/python/cpython/issues/100997)] [[PR](https://github.com/python/cpython/pull/100998)]
* ...

#### Related Links

* gh-85283
* https://discuss.python.org/t/a-new-c-api-for-extensions-that-need-runtime-global-locks/20668
* https://discuss.python.org/t/how-to-share-module-state-among-multiple-instances-of-an-extension-module/20663

</details>

### TODO

Here is the list of modules that need attention, in a *rough*, best-effort priority order.  Additional details (e.g. if there is an issue and/or PR) is found in the analysis table at the bottom.

* builtins (high priority)
   * [x] #101819
   * [x] **isolate** _collections
   * [x] #101520
   * [ ] #101509
* essential (higher priority)
   * [x] **port** _socket
   * [x] **port** readline (leave as is: only available in main interp)
   * [x] **isolate** _ssl
   * [x] **port** _pickle
   * [x] #117398
   * [x] **isolate** _asyncio
   * [x] #106078 
   * [x] **isolate** array
   * [x] **port** _ctypes
     * [x] #114314
     * [x] #117142
   * [x] **port** _tkinter (leave as is: only available in main interp)
   * [x] #103583
   * [x] #101714
   * [x] **isolate** _curses_panel (leave as is: only available in main interp)
   * [x] **isolate** _elementtree
   * [x] **isolate** pyexpat
   * [x] **port** winreg (Windows)
   * [x] **port** msvcrt (Windows)
* non-essential (lower priority)
   * [x] **port** winsound (Windows)
   * [x] **isolate** _lsprof

The above does not include test modules.  They don't need to be ported/isolated (except for a few which already have been).

---------------

### Modules Analysis

| module | builtin | Windows | PEP 594 | issue | PR | ported | # static types | # other global objects | # other globals | 
| --- | :---: | :---: | :---: | --- | --- | :---: | ---: | ---: | ---: |
| _asyncio |  |  |  |  |  | yes |  | [2](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L463-L464) |  |
| _collections  | X |  |  | (???) | (branch) | yes | [7](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L314-L318) |  |  |
| _ctypes |  |  |  |  |  | **\*\*NO\*\*** | [37](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L344-L380) | [6](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L419-L465) | [4](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L484-L530) |
| _curses |  |  |  |  |  | **\*\*NO\*\*** | [1](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L381) | [2](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L420-L442) | [4](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L486-L489) |
| _curses_panel |  |  |  |  |  | yes |  |  | [1](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L531) |
| _datetime |  |  |  | gh-71587 | gh-102995 | **\*\*NO\*\*** | [7](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L382-L388) | [10](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L443-L452) | [1](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L483) |
| _decimal |  |  |  |  |  | **\*\*NO\*\*** | [4](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L389-L392) | [10](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L411-L459) | [6](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L490-L495) |
| _elementtree |  |  |  |  |  | yes |  |  | [1](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L496) |
| _io | X |  |  | gh-101819 | gh-101520 | **\*\*NO\*\*** | [5](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L319-L325) |  |  |
| _lsprof |  |  |  |  |  | yes |  |  | [2](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L549-L550) |
| _multibytecodec |  |  |  |  |  | yes |  |  | [23](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L497-L519) |
| _pickle |  |  |  | (???) | (yes) | **\*\*NO\*\*** | [5](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L382-L388) |  |  |
| _socket |  |  |  |  |  | **\*\*NO\*\*** | [1](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L400) | [2](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L424-L425) | [3](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L523-L551) |
| _ssl |  |  |  | (???) | (branch) | yes |  |  | [1](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L532) |
| _tkinter |  |  |  |  |  | **\*\*NO\*\*** |  | [8](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L413-L469) | [9](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L533-L541) |
| _tracemalloc | X |  |  | gh-101520 |  | **\*\*NO\*\*** |  | [6](https://github.com/python/cpython/blob/main/Include/internal/pycore_tracemalloc.h#L86-L104) | [7](https://github.com/python/cpython/blob/main/Include/internal/pycore_tracemalloc.h#L68-L85) |
| array |  |  |  |  |  | yes |  | [1](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L460) |  |
| faulthandler  | X |  |  | gh-101509 |  | yes |  | [3+](https://github.com/python/cpython/blob/main/Include/internal/pycore_faulthandler.h#L39-L79) | [~22+](https://github.com/python/cpython/blob/main/Include/internal/pycore_faulthandler.h#L38-L84) |
| msvcrt |  | Y |  |  |  | **\*\*NO\*\*** | ??? | ??? | ??? |
| pyexpat |  |  |  |  |  | yes |  |  | [1](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L480) |
| readline |  |  |  |  |  | **\*\*NO\*\*** |  |  | [9](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L520-L548) |
| winreg |  | Y |  |  |  | **\*\*NO\*\*** | ??? | ??? | ??? |
| winsound |  | Y |  |  |  | **\*\*NO\*\*** | ??? | ??? | ??? |

<details>
<summary>test/example modules</summary>

These can be ported/isolated but don't have to be.  They are the lowest priority.

| module | issue | PR | ported | # static types | # other global objects | # other globals | 
| --- | --- | --- | :---: | ---: | ---: | ---: |
| xxmodule |  |  |  | [3](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L401-L403) | [1](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L427) |  |
| xxsubtype |  |  |  | [2](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L404-L405) |  |  |
| xxlimited_35 |  |  |  |  | [2](https://github.com/python/cpython/blob/main/Tools/c-analyzer/cpython/globals-to-fix.tsv#L416-L426) |  |
| ... |  |  |  |  |  |  |
| ... |  |  |  |  |  |  |

</details>

### Linked PRs
* gh-103093
* gh-102982
* gh-103094
* gh-103108
* gh-103248
* gh-103249
* gh-103250
* gh-103381
* gh-103540
* gh-103612
* gh-103893
* gh-103932
* gh-104020
* gh-104196
* gh-104506
* gh-104561
* gh-104725
* gh-113434
* gh-113555
* gh-113620
* gh-113630
* gh-113727
* gh-113774
* gh-113857
* gh-115130
* gh-115242
* gh-115301
* gh-23091
* gh-133674
* gh-133695

**Comment by ericsnowcurrently** [comment]
Ideally we'll get most of this done for 3.12.  FWIW, there isn't enough time if it's just me, so consider this a call out to whoever can help. 😄  Feel free to adjust the TODO list and (hidden) "analysis" table above.  Also feel free to reach out to any others that might be interested in pitching in here.  Thanks!!!

@erlend-aasland, @corona10, @kumaraditya303, etc.

**Comment by ericsnowcurrently** [comment]
@terryjreedy (for _tkinter)

**Comment by erlend-aasland** [comment]
I've got WIP branches for datetime, io, pickle, collections, and ssl. The three former are up as draft PRs.

(FYI, I have quite a bit of CPython time scheduled for next week.)

**Comment by ericsnowcurrently** [comment]
@markshannon

**Comment by erlend-aasland** [comment]
Regarding the freelists in `_asyncio`, Kumar and Guido had some thoughts over at https://github.com/python/cpython/issues/91375#issuecomment-1401757177.

**Comment by erlend-aasland** [comment]
Elementtree was isolated/ported in gh-92123.

**Comment by erlend-aasland** [comment]
`_lsprof` is also done, AFAICS.

**Comment by ericsnowcurrently** [comment]
Thanks for the updates!

**Comment by terryjreedy** [comment]
@serhiy-storchaka for _tkinter.  I have nothing to do with it.

**Comment by erlend-aasland** [comment]
AFAICS, `array` should be fine; `array_reconstructor` is in the module state already.

**Comment by corona10** [comment]
I will try to find which modules that I can support for this issue :)

**Comment by CharlieZhao95** [comment]
I tried submitting some PRs([Port _datatime module](https://github.com/python/cpython/pull/30522), [Convert _pickle module to use heap types](https://github.com/python/cpython/pull/30884)) for multi-phase initialization last year (delayed due to PEP 687 not being accepted at that time). 
Maybe now that I can start something new! Can I try to work for `_decimal` module, if no one has claimed it yet.

**Comment by shihai1991** [comment]
> I tried submitting some PRs([Port _datatime module](https://github.com/python/cpython/pull/30522), [Convert _pickle module to use heap types](https://github.com/python/cpython/pull/30884)) for multi-phase initialization last year (delayed due to PEP 687 not being accepted at that time).

Looks like you can reopen your PRs :)

**Comment by erlend-aasland** [comment]
> Maybe now that I can start something new! Can I try to work for `_decimal` module, if no one has claimed it yet.

Sure you can! I started doing some work on `_decimal` some weeks ago, but it is very far from complete; feel free to start with that branch :) https://github.com/erlend-aasland/cpython/tree/isolate-decimal

**Comment by kumaraditya303** [comment]
I volunteer to review the PRs as I have been doing, and leave `_asyncio` for me to handle.

**Comment by ericsnowcurrently** [comment]
FYI, I asked @Yhg1s (as 3.12 release manager) for his disposition on what changes are acceptable after beta 1, since there's certainly a chance we won't be able to finish with the above list in time.  He gave me the following guidance related to the work here on extension modules:

* no changes in module semantics after beta 1
* no non-internal C-API added after beta 1
* internal refactoring without semantics changes are okay after beta 1 but should be wrapped up ASAP so it all has time to bake
* as usual, bug fixes are completely okay after beta 1

Feel free to clarify/correct my summary, Thomas. 😄

With the above in mind, I'd like to be sure we have the proper priority for the modules that are left to port/isolate.  To that end, here are some questions:

* which modules are the most likely to require any change in semantics (regardless of how small)?
* is there any aspect of porting modules to multi-phase init that changes module semantics?  (I'm not aware of any.)
* are there any modules that will need any new non-internal API?  (I'm guessing there aren't.)

Again, ideally we will be able to wrap up this work in time for beta 1 (May 8).

**Comment by erlend-aasland** [comment]
For `readline`, I'm wondering if we should use a similar strategy as with `signal` and `faulthandler`. Thoughts?

**Comment by aisk** [comment]
I had took a look for `_ssl`, the only global state is a lock to prevent concurrency writes on log files, and according to its document: https://github.com/python/cpython/blob/8317d51996e68d8bb205385c1d47a9edcd128e7b/Modules/_ssl/debughelpers.c#L131-L136, the lock should be per process. Make it per interpreter will leads to concurrency issue.

Should we just leave it as it is?

**Comment by ericsnowcurrently** [comment]
We're probably okay leaving `_ssl` as it is then, for now.  The risk is in a race on creation of the lock, which is quite unlikely but still possible.

FWIW, this is related to https://discuss.python.org/t/a-new-c-api-for-extensions-that-need-runtime-global-locks/20668 (and the more general https://discuss.python.org/t/how-to-share-module-state-among-multiple-instances-of-an-extension-module/20663).

**Comment by ericsnowcurrently** [comment]
IMHO, it isn't critical.  The bigger priority is interpreter isolation for 3.12.  That said, do you think porting it to Argument Clinic needs to be done first in order to do the isolation work?

**Comment by arhadthedev** [comment]
> That said, do you think porting it to Argument Clinic needs to be done first in order to do the isolation work?

No, so I agree to move it into 3.13.

**Comment by erlend-aasland** [comment]
FTR, I'm working on multibytecodec today. PR coming tomorrow.

**Comment by ericsnowcurrently** [comment]
@nanjekyejoannah

**Comment by erlend-aasland** [comment]
I started with some low-hanging fruit on ctypes the other day. Kumar just reviewed it, and I merged the first part today. Now, for the next PRs, ctypes gets really interesting. There's some interesting metaclass/dict/baseclass hacks there. I have a draft for part 2 up (gh-103932), trying to use the same approach as we used for the defaultdict type when we isolated the collections module a few weeks ago, but I need some more eyes on it, so if you have time, please chime in :)

**Comment by ericsnowcurrently** [comment]
@erlend-aasland, @arhadthedev, @corona10, and anyone else porting/isolating extensions...

First of all, thanks for all the great work!  It's a big deal. ❤️

Second, what's the current status of the remaining modules.  If possible, please update the checklist and table in [the original post](https://github.com/python/cpython/issues/103092#issue-1644462938).

**Comment by ericsnowcurrently** [comment]
To anyone porting/isolating modules, I've added (in gh-104205) a module def slot to use to indicate support for per-interpreter GIL.  You'll add the following entry to the module's `PyModuleDef.m_slots`:

`{Py_mod_multiple_interpreters, Py_MOD_PER_INTERPRETER_GIL_SUPPORTED}`

I've added it to all the already isolated modules I'm aware of, and left notes for each of the modules that's been ported but still isn't fully isolated.

**Comment by erlend-aasland** [comment]
FTR, Kumar, Victor and I have been making great progress over at #101948 (#101819) lately. We'll probably be able to land the last PR for isolating \_io today (or at least, in the very near future).

**Comment by kumaraditya303** [comment]
`_io` is ported now!

**Comment by erlend-aasland** [comment]
What's missing with pyexpat/elementtree?

**Comment by brandtbucher** [comment]
I suggest we leave `readline` alone. I've been struggling with trying to port/isolate it for a while now, and honestly, it's just too... global.

For context: `readline` interacts with the global state of the wrapped GNU readline library, and most of its functionality is exposed via a registered `PyOS_ReadlineFunctionPointer` callback that has no way of recovering the module's state. Whenever any of the hooks are called, they need to get both C data *and Python objects* out of thin air.

So it really doesn't make any sense for it to be loaded multiple times in the same interpreter, and it's dangerous to load it in a subinterpreter. Since that's the case, I think it's totally fine (maybe even desirable) to leave it as single-phase init... especially since it's not even present in all builds, so most users need to be prepared for an `ImportError` anyways.

`tkinter` seems to be sort of in the same boat. For more context on the problematic `PyOS_InputHook` and `PyOS_ReadlineFunctionPointer` APIs, see #104668.

**Comment by erlend-aasland** [comment]
We _could_ solve them using _interpreter state_ and locking, but I'm not sure it is worth it. Also, I'm not sure there's a real use case for `readline` or `tkinter` and subinterpreters.

**Comment by kumaraditya303** [comment]
The beta freeze is tomorrow and the important modules are already ported over. The rest can be done in 3.13, we should maybe add a whatsnew entry for this, that `so so modules are isolated and are compatible with subinterpreters something something`.

**Comment by ericsnowcurrently** [comment]
For 3.12, we should still try to get the following done:

* _tracemalloc
* faulthandler
* _datetime
* _decimal
* _ctypes

Note that we do have until beta 2.

FWIW, I'm fine with leaving _tkinter, readline, and _curses/_curses_panel alone, since to me they only make sense in the main thread of the main interpreter.

**Comment by brandtbucher** [comment]
> Note that we do have until beta 2.

Just a heads-up, this is only one week away...

**Comment by brandtbucher** [comment]
> We could solve them using interpreter state and locking, but I'm not sure it is worth it. Also, I'm not sure there's a real use case for readline or tkinter and subinterpreters.

Yeah, I think that *if* this is worth fixing, it's probably just easier to introduce a better API that third-party extensions can use, rather than using the interpreter state as a dumping ground for `readline`/`tkinter` state.

**Comment by arhadthedev** [comment]
3.13 development launched so *Deprecated (lower priority)* modules are to be gone in gh-104773. I removed them from the issue list and the *Modules Analysis* table.

**Comment by erlend-aasland** [comment]
FTR, `_datetime` is almost ready. I've got a wip branch for `_ctypes`, but there's lots left there.

**Comment by CharlieZhao95** [comment]
The [`_decimal`](https://github.com/python/cpython/pull/103381) module is also nearing completion, it currently passes all test cases. But I believe it's still not final, if anyone is free please help to review it. Thanks to all!

**Comment by erlend-aasland** [comment]
I recon we want to leave `_tkinter` as is, for the same reasons as for `readline` and `_curses`. cc. @ericsnowcurrently

**Comment by ericsnowcurrently** [comment]
makes sense

**Comment by vstinner** [comment]
Statistics on C types in the main branch:

* 78% (207/266) are implemented as heap types
* 22% (59/266) are implemented as static types

```
$ grep -E 'static PyTypeObject .* =' $(find -name "*.c"|grep -v Doc/)|wc -l
59
$ grep -E 'PyType_Spec .* =' $(find -name "*.c"|grep -v Doc/)|wc -l
207
```

Statistics on C extensions in the main branch:

* 85% (120/141) C extensions use the multi-phase initialization API (PEP 489), and likely use heap types
* 15% (21/141) C extensions use the legacy module API, and likely use static types

```
vstinner@mona$ grep 'PyModuleDef_Init *(' $(find -name "*.c"|grep -v Doc/)|wc -l
120
vstinner@mona$ grep 'PyModule_Create *(' $(find -name "*.c"|grep -v Doc/)|wc -l
21
```

**Comment by neonene** [comment]
@erlend-aasland   I'd like `_datetime` static types to be heap types to help #113601. Would you mind if I open an issue and take the type conversion part from your POC (without putting them into some states)?

**Comment by vstinner** [comment]
> @erlend-aasland I'd like _datetime static types to be heap types to help https://github.com/python/cpython/pull/113601. Would you mind if I open an issue and take the type conversion part from your POC (without putting them into some states)?

Converting static types to heap types is not the blocker issue for the datetime. The problem of datetime is its C API.

It reminds me the problem of the AST C API. We tried different things but we had multiple crashes. At the end, I moved AST types to the PyInterpreterState, they are initialized at the first access. This way, from any code path, you always meet the same types. Before, depending on how types were created, you might meet different types and so PyAST_Check() returned false which was very surprising.

I can help you if you open an issue. But please, first, move definition of these types to PyInterpreterState (put `PyTypeObject*` pointers there), in a new "datetime" state even if it's shared library.

**Comment by neonene** [comment]
Maybe already known, but reloading any module (e.g.`-R`) seems to be possibly unsafe:
```py
import sys
from _decimal import Decimal as D1
del sys.modules['_decimal']
from _decimal import Decimal as D2

print(a := D1(1), b := D1(1), a == b, a is b)  # load
print(c := D2(1), d := D2(1), c == d, c is d)  # reload
print(a == c)
```

* 3.12: Static `Decimal` type
```py
1 1 True False
1 1 True False
True
```

* 3.13: Heap `Decimal` type
```py
1 1 True False
1 1 True False
False   # a and c have different module states
```
Indeed, `PyInterpreterState` should be effective for `_datetime` to be used in main and isolated sub-interpreters.

**Comment by encukou** [comment]
What do you mean by *unsafe*? If something's unsafe at the C level (e.g. undefined behaviour, crashes), that's a bug.

This behaviour is known; see https://docs.python.org/3/howto/isolating-extensions.html#surprising-edge-cases and the warning in https://docs.python.org/3/library/sys.html#sys.modules

**Comment by neonene** [comment]
I've opened a `_datetime` issue: #117398

**Comment by ericsnowcurrently** [comment]
@Yhg1s, could we get an exception for 3.13 like we got for 3.12, regarding porting stdlib extensions to multi-phase init?  You let us land those changes up to beta 2.

This time around there are only two modules left to be done, and the PRs for both are close enough that I expect they could be wrapped up before beta 2 (end of May).

**Comment by neonene** [comment]
@vstinner

From https://github.com/python/cpython/pull/118357#issuecomment-2093208624:
>Which crash? This PR is related to issue https://github.com/python/cpython/issues/117398 which doesn't mention any crash.

I said at https://github.com/python/cpython/issues/103092#issuecomment-2024497734 in this thread:
> I'd like `_datetime` static types to be heap types to help #113601. 

You replied:
> I can help you if you open an issue.

Then, I opened the issue #117398.  I think we are aware of #113601, whose original issue was raised a year ago and you said there (https://github.com/python/cpython/issues/100911#issuecomment-1497939136):

> It will work as expected once more extensions will be converted to the multi-phase initialization API.

We know well that the porting resolves existing different issues derived from static types and single-phase init.  My volunteer is following the generic approach that seems to be supported by active core devs, especially by you.  Also, the recent crash case you reported at #116467 on `_ctypes` (actually `ctypes`)  was actually fixed by multi-phase init.

@kumaraditya303 has not been in good health, right?  And @erlend-aasland appears to be not working individually on this project, and to be exhausted: https://github.com/python/cpython/issues/110415#issuecomment-1750242478.

As to `_datetime` C-API, there seems to be no technical reason to block the isolation. Even if it blocks, I do not understand why there was no attempt to unblock nor to investigate further.

>>> Such change is too late for beta1. It's better to wait for the beginning of the 3.14 dev cycle.
>>
>> How do you think of the situation?
>
> In my experience, such change is likely to cause regression and need more time to be stabilized.

Again, how do you think of the delay?  Maybe isolated `_datetime` could be stabilized even at this point if we tried to work without bias nor reluctance?  Why we cannot stabilize `_datetime` in the regular way with many eye-balls on 3.13?

**Comment by vstinner** [comment]
Please open a dedicated issue for the datetime crash.

**Comment by vstinner** [comment]
> @Yhg1s, could we get an exception for 3.13 like we got for 3.12, regarding porting stdlib extensions to multi-phase init? You let us land those changes up to beta 2.

I don't think that it's a good idea. Regressions are complex  to debug/fix, and it's better to continue the work in Python 3.14 and learn how to work around Python 3.13 limited isolation.

**Comment by Yhg1s** [comment]
> @Yhg1s, could we get an exception for 3.13 like we got for 3.12, regarding porting stdlib extensions to multi-phase init? You let us land those changes up to beta 2.
> 
> This time around there are only two modules left to be done, and the PRs for both are close enough that I expect they could be wrapped up before beta 2 (end of May).

Which modules are we talking about? You referenced this from a _datetime change that didn't seem to actually switch to multi-phase init.

**Comment by neonene** [comment]
> Please open a dedicated issue for the datetime crash.

Opened #118608, whose title is taken from the `_ctypes` crash issue.

**Comment by vstinner** [comment]
> Opened https://github.com/python/cpython/issues/118608, whose title is taken from the _ctypes crash issue.

Oh, Python 3.12 is also affected. We cannot isolate datetime in 3.12 to fix the issue. We will need a different fix for 3.12. So 3.13 can use the same fix.

**Comment by neonene** [comment]
> We will need a different fix for 3.12.

Unfortunately, `ctypes` also crashes on 3.12:
* #113576

**Comment by ericsnowcurrently** [comment]
> Which modules are we talking about? You referenced this from a _datetime change that didn't seem to actually switch to multi-phase init.

datetime and ctypes.  Those are the only two left from the list at the top of this issue.

**Comment by kumaraditya303** [comment]
`_datetime` and `_ctypes` have been ported so this is completed now.
